### ISSUE TITLE
EventSource events not being shown as Traces in Azure Portal

### ISSUE THREAD

**Issue by eparizzi** [Microsoft/ApplicationInsights-dotnet-logging#262]
I have a simple .NET Framework 4.6.1 console application with the following `ApplicationInsights.config`

```xml
<?xml version="1.0" encoding="utf-8"?>
<ApplicationInsights xmlns="http://schemas.microsoft.com/ApplicationInsights/2013/Settings">
  <InstrumentationKey>8c738a10-bf42-49f3-95f9-369f982f7716</InstrumentationKey>
  <TelemetryModules>
    <Add Type="Microsoft.ApplicationInsights.EventSourceListener.EventSourceTelemetryModule, Microsoft.ApplicationInsights.EventSourceListener">
      <Sources>
        <Add Name="MyCompany" Level="LogAlways" />
      </Sources>
    </Add>
  </TelemetryModules>
</ApplicationInsights>
```

Then I have a simple EventSource class as follows:
```csharp
using System;
using System.Diagnostics;
using System.Diagnostics.Tracing;

namespace MyNamespace
{
    [EventSource(Name = "MyCompany")]
    public class LogEventTrace : EventSource
    {
        public LogEventTrace()
        {
        }

        [NonEvent]
        public void Error(string message)
        {
            ErrorInternal(message, string.Empty, string.Empty);
        }

        [NonEvent]
        public void Error(Exception exception, string message)
        {
            ErrorInternal(message, exception.Message, exception.ToString());
        }

        [NonEvent]
        public void Critical(string message)
        {
            CriticalInternal(message);
        }

        [NonEvent]
        public void Warning(string message)
        {
            WarningInternal(message);
        }

        [NonEvent]
        public void Information(string message)
        {
            InformationInternal(message);
        }

        [NonEvent]
        public void Verbose(string message)
        {
            VerboseInternal(message);
        }

        [Event((int)TraceEventType.Critical)]
        private void CriticalInternal(string message)
        {
            WriteEvent((int)TraceEventType.Critical, message);
        }

        [Event((int)TraceEventType.Warning)]
        private void WarningInternal(string message)
        {
            WriteEvent((int)TraceEventType.Warning, message);
        }

        [Event((int)TraceEventType.Information)]
        private void InformationInternal(string message)
        {
            WriteEvent((int)TraceEventType.Information, message);
        }

        [Event((int)TraceEventType.Verbose)]
        private void VerboseInternal(string message)
        {
            WriteEvent((int)TraceEventType.Verbose, message);
        }

        [Event((int)TraceEventType.Error)]
        private void ErrorInternal(string message, string exceptionMessage, string exceptionDetails)
        {
            WriteEvent((int)TraceEventType.Error, message, exceptionMessage, exceptionDetails);
        }
    }
}
```

In my main method I create a new instance of the `LogEventTrace` class and call some of the methods to write events. 
I can see in the output window one line per each call like

Application Insights Telemetry: 
```json
{"name":"Microsoft.ApplicationInsights.XXXXXXXXXXXXXXXXXXX.Message","time":"2019-02-13T19:39:56.5661395Z","iKey":"XXXXXXXXXXXXXXXXXXXX","tags":{"ai.cloud.roleInstance":"EMO-DESKTOP","ai.internal.sdkVersion":"evl:2.8.1-23827"},"data":{"baseType":"MessageData","baseData":{"ver":2,"message":"n/a","severityLevel":"Information","properties":{"Task":"0x0000FFF6","Channel":"None","ProviderGuid":"4380a117-df11-523f-e338-737b1cb0925e","ProviderName":"MyCompany","EventId":"8","Opcode":"Info","Keywords":"0x0000F00000000000","message":"Some test message.","EventName":"InformationInternal"}}}}
```

So apparently, ApplicationInsights is correctly picking the EventSource events, right?
But I don't see them in the Azure Portal > Application Insight resouce > Search.

It's worth to mention that manually instantiating a `TelemetryClient` and writing via the `TrackTrace()` method works well.

**Comment by karolz-ms** [comment]
Emiliano, could you share a full example that exhibits the problem?

From a cursory look it does seem like your code should work. One thing though: the way you use `TraceEventType` enumeration as a way to assign event IDs... you do realize you can have many more events in a single `EventSource` than there are values for `TraceEventType`?

https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventattribute.-ctor?view=netcore-2.2

**Comment by eparizzi** [comment]
@karolz-ms Yes, I see that. It's not my class but I see your point. Although I don't think is relevant to the problem, right?

I'll make a repo to reproduce the issue.

**Comment by karolz-ms** [comment]
@eparizzi correct, the way you assign event IDs should not affect the issue in question

**Comment by eparizzi** [comment]
@karolz-ms I couldn't reproduce the issue in a fresh console app.

But on my current project where the issue happens, if I do a debug session, then expand the ApplicationInsights.config and open the "Search debug session telemetry", I can see all the EventSource events there properly (and also in the Output window as I step through the code as I mentioned before).
But they do not reach Azure.

What can be the reasons for this?
Does the project needs to wait a bit before being terminated for the events to be flushed out to Azure as  the example in the docs do when using the TelemetryClient directly and using .Flush() ?

**Comment by karolz-ms** [comment]
Sounds like a problem with the telemetry channel/its configuration. I am not an expert in this area but @cijothomas and @Dmitry-Matveev are.

**Comment by Dmitry-Matveev** [comment]
@eparizzi , you're correct. Default telemetry batching/submission is 30 seconds or ~500-1000 items.
So, if the application exits immediately, it does not send telemetry out. If you are using `InMemoryChannel` (default if not ApplicationInsights.config) is present - `Flush()` will be "syncronous enough" to use it as is. If you leverage `ServerTelemetryChannel`, `Flush()` does not actually post to the AI endpoint but rather tries to save to local buffer/storage to deliver later on (which won't happen when app exits right away).

**Comment by eparizzi** [comment]
@Dmitry-Matveev Is that somewhere in the docs???

This is a console app with an ApplicationInsights.config which does some work and then exits. In the process it logs some events using EventSource which AI is configured to pick up.
The process takes less than 30s and it doesn't produce 500 events.

What am I supposed to do? Put a Task.Delay(300000) before terminating the process?

**Comment by Dmitry-Matveev** [comment]
In this case, i'd simply recommend using `InMemoryChannel` and use `Flush()` without any delay before the app exits. [Here](https://docs.microsoft.com/en-us/azure/azure-monitor/app/windows-desktop#telemetry) is the relevant doc (i think you may skip Sleep/Delay, though..)

**Comment by eparizzi** [comment]
Ok, but the thing is that I'm not using the `TelemetryClient` directly, I'm just writing events using the `EventSource `class and those are automagically picked up by the [Microsoft.ApplicationInsights.EventSourceListener](https://www.nuget.org/packages/Microsoft.ApplicationInsights.EventSourceListener) package.

**Comment by Dmitry-Matveev** [comment]
I see. @cijothomas / @karolz-ms , any way to get to the instance of the Channel or Telemetry Client in case of `EventSourceListener` to Flush it? I guess something like `TelemetryConfiguration.Active.TelemetryChannel.Flush()` if Active is used by EventSource module?

Alternatively, you can provide your own configuration to the EventSourceModule at construction time (but it will require configuring this module in code

@eparizzi , unrotunately, `PersistenceChannel` is deprecated, documentation is pending an update to reflect this.

**Comment by eparizzi** [comment]
Calling `TelemetryConfiguration.Active.TelemetryChannel.Flush();` before terminating the process seemed to work. No need to wait 30s!
